### PR TITLE
feat(chat): Attachment visualizer

### LINKED ISSUE #524: task(chat): embeds images inline in chat
Image

Currently all attachments view in the "no preview available" view, instead videos and pictures should be rendered inline.

### PR DESCRIPTION
**Pull request by Flemmli97**
### What this PR does 📖

- Makes it so attachments like png etc. will show in chat as image instead of just having the file name

~~- This is the draft for having attachment embeds in chat. Will show a preview of the attachment of messages~~
~~- Currently due to warp not providing any thumbnail/picture for files it wont do anything~~
~~- Current status of the PR includes the ui part such that if warp does provide the missing things it should work out of the box. It uses `warp::constellation::file::File#thumbnail` which is currently never set by warp to obtain a path to the file thumbnail.~~
~~- To test the current impl one can replace the thumbnail value in `kit::components::file_embed::mod.rs#68` with the following to use the current profile pic:~~
```rust
    let state = use_shared_state::<State>(cx)?;
    let thumbnail = state.read().get_own_identity().profile_picture();
```

### Which issue(s) this PR fixes 🔨

- Resolve #524

### CODE REVIEW

**Comment by stavares843** [comment]
**AI CODE REVIEW**

Feedback on code changes:

Patch 1/5:

The code changes in mod.rs file of the file_embed component and the associated style.scss files seem fine.
The changes include adding a new thumbnail prop to the Props struct, handling the thumbnail prop in the component rendering logic, and updating the associated styles.
The code appears to do what it says in the commit message, which is adding attachment rendering functionality.
Patch 2/5:

The code changes in mod.rs and style.scss files of the file_embed component and mod.rs and style.scss files of the message and message_reply components seem reasonable.
The changes include adding a new big prop to the Props struct of the file_embed component, updating the rendering logic to handle the big prop, and adjusting styles accordingly.
The code appears to make the attachment embed look nicer by providing an option to display a larger image if the big prop is set to true.
Patch 3/5:

The code change in mod.rs file of the file_embed component looks fine.
The change removes an unused import, which is a good cleanup.
The code appears to address a Clippy warning or unused code.
Patch 4/5:

The code changes in mod.rs files of the message, message_reply, and chatbar components seem fine.
The changes include adding the thumbnail prop to the Props struct of the message component and handling it in the component's rendering logic.
The code appears to introduce the use of the thumbnail for files in messages.
Overall, the code changes in the patches seem reasonable and align with their respective commit messages. No bugs or major issues are found.

File: kit/src/components/file_embed/mod.rs [hunk 1]
```diff
@@ -28,6 +28,11 @@ pub struct Props<'a> {
     // used for the button. defaults to a download icon
     button_icon: Option<Icon>,
 
+    // The thumbnail for the file. If existent
+    thumbnail: Option<String>,
+
+    big: Option<bool>,
```
Comment (dariusc93) [inline kit/src/components/file_embed/mod.rs:34]: Would this be better to just be a boolean instead?

**Review by dariusc93 — APPROVED** [review]

**Comment by stavares843** [comment]
**AI CODE REVIEW**

Feedback on the code changes:

Overall, the code changes look good and seem to align with the commit messages.
In the FileEmbed component in mod.rs, the changes introduce a new thumbnail property, which is optional. The code handles the presence of a thumbnail and renders it accordingly.
There are some style changes in the SCSS files, primarily related to the file_embed and message components.
In file_embed/style.scss, there is a new .image-container class introduced, which positions the image and download button correctly.
In file_embed/mod.rs, the big property is added to the Props struct and used to determine if the file embed should be rendered with a larger size. It is passed down to the FileEmbed component and used to conditionally apply the big class in the rendered HTML.
In message/mod.rs, the thumbnail property is passed down to the FileEmbed component, allowing it to display the thumbnail of the file.
In message_reply/style.scss, there are some minor style changes related to alignment.
In file_embed/mod.rs, the import of common::state::State is no longer needed and can be removed.

**Comment by stavares843** [comment]
In file_embed/mod.rs, the import of common::state::State is no longer needed and can be removed.

is that accurate?

**Review by stavares843 — COMMENTED** [review]
https://github.com/Satellite-im/Uplink/assets/29093946/aa2147d0-7647-435e-984c-98e5f2856689

is it possible to improve the size of the preview? atm is somewhat big, even when sending very small files like favicons always display the same size

macOS, m1

**Review by stavares843 — CHANGES_REQUESTED** [review]
sending specific images crash's


crash log
https://gist.github.com/stavares843/35e8cf30f331acba46ae6c6c9669bb83

macOS, m1

image (zipped)

[2B986EBC-A8C4-4C68-BF5D-0F5CBFB1ACB1.png.zip](https://github.com/Satellite-im/Uplink/files/11522667/2B986EBC-A8C4-4C68-BF5D-0F5CBFB1ACB1.png.zip)

flow
- unzip
- drag to chat
- send

the crash happens after sending, display the loading and then before displaying the image crash's

doesn't crash on dev because theres no image preview in chat in dev

**Comment by dariusc93** [comment]
> Gravacao.do.ecra.2023-05-20.as.16.34.29.mov
> 
> is it possible to improve the size of the preview? atm is somewhat big, even when sending very small files like favicons always display the same size
> 
> macOS, m1

That will be something ill handle on my end in a different update on warp (notice this while testing this branch).

**Comment by dariusc93** [comment]
> sending specific images crash's
> 
> crash log https://gist.github.com/stavares843/35e8cf30f331acba46ae6c6c9669bb83
> 
> macOS, m1
> 
> image (zipped)
> 
> [2B986EBC-A8C4-4C68-BF5D-0F5CBFB1ACB1.png.zip](https://github.com/Satellite-im/Uplink/files/11522667/2B986EBC-A8C4-4C68-BF5D-0F5CBFB1ACB1.png.zip)
> 
> flow
> 
>     * unzip
> 
>     * drag to chat
> 
>     * send
> 
> 
> the crash happens after sending, display the loading and then before displaying the image crash's
> 
> doesn't crash on dev because theres no image preview in chat in dev

I havent been able to reproduce this on my mac

**Comment by phillsatellite** [comment]
> > sending specific images crash's
> > crash log https://gist.github.com/stavares843/35e8cf30f331acba46ae6c6c9669bb83
> > macOS, m1
> > image (zipped)
> > [2B986EBC-A8C4-4C68-BF5D-0F5CBFB1ACB1.png.zip](https://github.com/Satellite-im/Uplink/files/11522667/2B986EBC-A8C4-4C68-BF5D-0F5CBFB1ACB1.png.zip)
> > flow
> > ```
> > * unzip
> > 
> > * drag to chat
> > 
> > * send
> > ```
> > 
> > 
> >     
> >       
> >     
> > 
> >       
> >     
> > 
> >     
> >   
> > the crash happens after sending, display the loading and then before displaying the image crash's
> > doesn't crash on dev because theres no image preview in chat in dev
> 
> I havent been able to reproduce this on my mac

I could not get it to crash but I could not get attachment to appear when sent 

Mac m2 
OS 13.3.1 

2B986EBC-A8C4-4C68-BF5D-0F5CBFB1ACB1

**Comment by Flemmli97** [comment]
tested on linux. works for both attachments send and received

**Comment by Flemmli97** [comment]
tested on mac. same behaviour as phill. no thumbnail. the thumbnail field for files is empty.

**Comment by Flemmli97** [comment]
also regarding image size: images get clamped to a max of 700x700 px. smaller images will render smaller. currently it seems all thumbnail images from warp are 500x500

**Comment by dariusc93** [comment]
image

It appears fine for me, but i do know some images had issues being encoded, which would lead to some not showing up. May be more of issue with the library being used and would require additional work for those set of images, which can come a bit later. 

As for the dimensions, in warp its set for 128x128 for thumbnails, but i increased it to 500x500 in uplink. I did forget to add a check in for the dimensions before resizing so that will be added in a shortly.

**Comment by Flemmli97** [comment]
it was just a plain png i used.
on mac air. 12.6.4

**Comment by dariusc93** [comment]
> it was just a plain png i used. on mac air. 12.6.4

If it was similar to what @stavares843 was using, then i will probably also work in a workaround for that. I think its good to merge it in though and i can bring the changes over to correct a couple of things shortly.

**Comment by stavares843** [comment]
sounds good 🔨

**Comment by stavares843** [comment]
**AI CODE REVIEW**


Review of code changes:

Commit 1/5:

Adds a new prop thumbnail to the Props struct in file_embed/mod.rs.
Adds support for rendering a thumbnail image if it exists.
Changes the layout of the file embed based on the presence of a thumbnail.
Fixes an issue where the with_download_button prop was not being used correctly.
The code changes appear to align with the commit message.
Commit 2/5:

Makes improvements to the styling of the file embed component.
Adds support for a larger image display with the addition of the big prop.
Modifies the CSS rules for the image container and image elements.
Changes the maximum width and height of the image based on the big prop.
The code changes appear to align with the commit message.
Commit 3/5:

Removes an unused import statement in file_embed/mod.rs.
The code change appears to align with the commit message.
Commit 4/5:

Adds support for rendering the thumbnail of a file in the message component.
Modifies the Props struct in message/mod.rs to include the thumbnail prop.
Passes the thumbnail prop to the file embed component.
The code change appears to align with the commit message.
Overall, the code changes seem to implement the described features and improvements. No bugs or major issues are apparent.

**Comment by stavares843** [comment]
do we wanna address any other changes here?

**Comment by stavares843** [comment]
cc @Flemmli97 let us know if this is ready to do code review and testing

**Comment by Flemmli97** [comment]
> cc @Flemmli97 let us know if this is ready to do code review and testing

thought it was good to go according to darius

**Comment by stavares843** [comment]
awesome, thanks 🎉

**Review by stavares843 — APPROVED** [review]

**Pull request merged**